sdks/python/apache_beam/io/gcp/datastore/v1new/datastoreio.py

```diff
@@ -276,15 +277,33 @@ class _Mutate(PTransform):
   Only idempotent Datastore mutation operations (upsert and delete) are
   supported, as the commits are retried when failures occur.
   """
-  def __init__(self, mutate_fn):
+
+  # Default hint for the expected number of workers in the ramp-up throttling
+  # step for write or delete operations.
+  _DEFAULT_HINT_NUM_WORKERS = 500
```

Comment:
Do you know how this behaves when Dataflow Autoscaling is enabled ? Users can set two parameters num_workers and max_num_workers. We start with num_workers and Dataflow job may scale up to "max_num_workers".

Also, the sink reports the throttled time to Dataflow service which adjusts autoscaling decisions accordingly (i.e. do not scale up if sink is significantly throttled).

I'm wondering if statically setting the expected number of workers statically could result in a significant regression for existing users. 

cc: @nehsyc from autoscaling team.

Reply:
I've thought about this before. Generally what I would expect is a slow start, but autoscaling quickly scaling up when the throttling limit increases, yielding a similar result to starting with the desired number of workers.   
However, I was weighing whether to report `throttlingMs` or if that stifles autoscaling (the Firestore implementation reports it). I thought I remembered previous test showing not much of a difference, but I've run a few more tests and I do see a significant difference now. This is an example I ran with the Java implementation and `maxNumWorkers` at 50, where I commented out incrementing the `throttlingMs` counter: 
![Screen Shot 2021-05-21 at 11 35 15 AM](https://user-images.githubusercontent.com/76013657/119188283-4f347800-ba2f-11eb-851d-8102a3834230.png)

That's about what I would expect, autoscaling scales up after 10-15 minutes, then the ramp-up proceeds normally. I just ran that same example with the original implementation, reporting `throttlingMs`, and Dataflow seems prone to getting stuck at 1 worker with very little throughput.

So if I drop `throttlingMs`, autoscaling should be fairly normal in terms of behavior. Are there other side effects with autoscaling I should consider?

Reply:
Thanks Cham for looping me in. 

I am not too sure about the default num workers hint. My _intuition_ is that it could impact Dataflow autoscaling behavior in the following ways:

- default workers >> the actual workers a Dataflow job starts with
The ramp up process would be slow (probably much slower than necessary) and Dataflow may 
  - upscale the workers aggressively due to the slow progress if `throttling-msecs` is not reported. In your test, it was not a big problem since you set max number of workers, but imagine if there was no limit set, how many workers Dataflow might allocate.
  - or cap the worker pool size at a small number due to throttling if `throttling-msecs` is reported. In your another test, I would imagine that the rate limit was low (i.e., BUDGET / 500) even though there was only one worker, which might already trigger throttling signals and therefore limit upsizing to more than 1.
 
- default workers << the actual workers a Dataflow job starts with
Dataflow may 
  - upscale the workers moderately if `throttling-msecs` is not reported.
  - cap the workers but not by too much if `throttling-msecs` is reported.
 
  However, the mutation rate could increase faster than what's recommended in the best practice, possibly causing Firestore to be overwhelmed.

Ideally the per-worker budget is set by the runner given a global limit. I don't know that is already a thing in Beam. If we have to keep this default number of workers, can we make it a configurable option and override it with num_workers if supplied by users? Otherwise it might be safe to assume that a Dataflow job always start with a small number of (or a single) workers. `throttling-msecs` might still be needed to avoid excessive upsizing. But anyway these are just based on me thinking :p We should probably run more experiments to verify.

(Another idea is to keep the default num of workers hint and keep reporting `throttling-msecs` but in a reduced frequency. The purpose is to give Dataflow some chance to grow the num of workers but also let it be aware of some implicit rate limit. But I would not bother if a simpler solution can work.)

Reply:
That's a good point, simply keeping or dropping `throttling-msecs` might under- or overshoot the target... 

The hint is already configurable, and with the warning messages and ramp-up as a separate step it should be easy for the user to recognize the throttling and adjust the value. Even egregious misconfiguration becomes irrelevant after ~1 hour as the ramp-up is exponential, so expected impact is low.  
It's also possible to turn off ramp-up, though that is obviously not recommended. 

However, the dynamic worker count in case of autoscaling certainly complicates this. The crux of the issue is that we can't get an up-to-date worker count from the runner or Beam in general, so we're going by a (configurable) rough estimate. I've looked into auto-filling the value from `maxNumWorkers` if the runner is Dataflow, but that introduces a circular dependency at least in Java.  
All of this should be fine I think if autoscaling scales to the maximum in a reasonable amount of time as it did in my experiment, but I didn't consider the possibility of overscaling if there is no limit. My other experiment that reported `throttling-msecs` had `maxNumWorkers` as well as the hint set to 50, so the budget wasn't quite as large, but I did have basically 2.5 hours of just one worker. Weirdly though as I just found out, `throttling-msecs` actually started stagnating (as expected), but Dataflow still didn't scale until 2 hours later. Is there a long memory on that metric, and is that something that could be fixed by reducing reporting frequency?  
This is what this looked like on the "report throttling-msecs" experiment (note the timestamps): 
![image](https://user-images.githubusercontent.com/76013657/119402371-b3ef0d00-bc91-11eb-8ef9-ab7766814131.png)
I'd be happy to run more experiments or jump on GVC to explain my observations if you'd like.

Reply:
Bumping this because the edit to my previous comment is easy to miss. The gist is that reporting throttling-msecs does not break autoscaling as far as I can tell.  
I did most of my testing with the Java implementation. I just noticed that not only did I miss adding the throttling counter in the Python implementation, but it appears Datastore never had one for the adaptive throttler either. Does the metric work the same in Python, can I just add a counter called `throttling-msecs`?

Reply:
Thanks for bumping this! I indeed missed the edits above.

It also appears weird to me that Dataflow didn't upscale for two hours despite the plateaued throttling time. My best guesses would be that the workers didn't report much progress causing Dataflow to stick to the initial scaling of the job, or the low CPU utilization prevented Dataflow to upscale.

My previous concern about reporting throttling signals was mainly the slow ramp-up of number of workers especially when we started with very little budget for each worker. If that is not as concerning as I imagined then I vote for reporting `throttling-msecs`.

Most Dataflow pipelines start with a small number of workers (<10) so by setting the default value to 500 we almost always over-constrain the rate by a lot at the beginning. Maybe consider lowering it.

I am not familiar with how counters work in Python. Maybe @chamikaramj can comment more on that.

Reply:
I haven't been able to find a reference to throttling counters in the Python Dataflow runner code. I've added a counter with the same name as the others (`cumulativeThrottlingSeconds`) into the DoFn for potential future proofing., @chamikaramj let me know if that is alright.

I have done a few more tests with the default value of 500, and even then autoscaling quickly went up to whatever maximum I set (69 was max quota for my test account). We originally picked such a high default to make sure unconfigured one-off pipelines don't cause much overload even at 2000 workers, if worker count tops out at a much lower number there's always the option for a user to reconfigure.

Reply:
cumulativeThrottlingSeconds results in "total_throttler_wait_time_seconds" property of the Dataflow API being set which results in reporting the per-worker throttled number of seconds to the Dataflow service. (This part of the code, Dataflow Python legacy worker, is not open source).